substrate/frame/election-provider-multi-block/src/signed/tests.rs

```diff
@@ -633,4 +640,626 @@ mod e2e {
+			assert_eq!(Submissions::<Runtime>::submitters_count(current_round), 0);
+			assert!(!Submissions::<Runtime>::has_leader(current_round));
+		});
+	}
+
+	#[test]
+	fn after_rejecting_does_not_call_verifier_start_if_no_leader_exists() {
+		ExtBuilder::signed().build_and_execute(|| {
+			roll_to_signed_open();
+			assert_full_snapshot();
+
+			// Submit only an invalid solution
+			let score = ElectionScore { minimal_stake: 10, sum_stake: 10, sum_stake_squared: 100 };
+			assert_ok!(SignedPallet::register(RuntimeOrigin::signed(99), score));
+			assert_ok!(SignedPallet::submit_page(
+				RuntimeOrigin::signed(99),
+				0,
+				Some(Default::default())
+			));
+
+			// Verify we have exactly one submission
+			let current_round = SignedPallet::current_round();
+			assert_eq!(Submissions::<Runtime>::submitters_count(current_round), 1);
+			assert!(Submissions::<Runtime>::has_leader(current_round));
+
+			roll_to_signed_validation_open();
+
+			assert!(matches!(VerifierPallet::status(), crate::verifier::Status::Ongoing(_)));
+
+			roll_next(); // Block 1: processes page 2 (empty)
+			roll_next(); // Block 2: processes page 1  (empty)
+			roll_next(); // Block 3: processes page 0 (the only submitted page)
+
+			// After 3 blocks, the invalid solution should be processed and discarded
+			// Verify no-restart conditions are met
+			assert!(crate::Pallet::<Runtime>::current_phase().is_signed_validation());
+			assert!(!Submissions::<Runtime>::has_leader(current_round));
+			assert_eq!(Submissions::<Runtime>::submitters_count(current_round), 0);
+			assert_eq!(VerifierPallet::status(), crate::verifier::Status::Nothing);
+
+			// Check that expected events were emitted for the rejection
+			assert_eq!(
+				signed_events(),
+				vec![
+					Event::Registered(
+						0,
+						99,
+						ElectionScore { minimal_stake: 10, sum_stake: 10, sum_stake_squared: 100 }
+					),
+					Event::Stored(0, 99, 0),
+					Event::Slashed(0, 99, 6),
+				]
+			);
+			assert_eq!(
+				verifier_events(),
+				vec![
+					crate::verifier::Event::Verified(2, 0),
+					crate::verifier::Event::Verified(1, 0),
+					crate::verifier::Event::Verified(0, 0),
+					crate::verifier::Event::VerificationFailed(0, FeasibilityError::InvalidScore),
+				]
+			);
+		});
+	}
+
+	#[test]
+	fn after_accepting_one_solution_verifier_is_idle_if_no_leader_exists() {
+		ExtBuilder::signed().build_and_execute(|| {
+			roll_to_signed_open();
+			assert_full_snapshot();
+
+			// Submit only one good solution that will be accepted
+			let good_solution = mine_full_solution().unwrap();
+			load_signed_for_verification(99, good_solution.clone());
+
+			// Verify we have exactly one submission
+			let current_round = SignedPallet::current_round();
+			assert_eq!(Submissions::<Runtime>::submitters_count(current_round), 1);
+			assert!(Submissions::<Runtime>::has_leader(current_round));
+
+			// Move to verification phase
+			roll_to_signed_validation_open();
```

Comment:
since in this case you are focusing on only seeing what happens after a solution is verified, I would use a helper function that does all of this, and no longer (re) assert on every step, or call `roll_next()` 3 times and so on. I think we have helper functions that:

1. Load a signed solution
2. roll until it is verified